extensions/amp-analytics/0.1/iframe-transport-message-queue.js

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {
+  AMP_ANALYTICS_3P_EVENT_MESSAGES_TYPE,
+} from '../../../src/3p-analytics-common';
+import {SubscriptionApi} from '../../../src/iframe-helper';
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.IframeTransportMessageQueue';
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ */
+export class IframeTransportMessageQueue {
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!HTMLIFrameElement} frame The cross-domain iframe to send
+   * messages to
+   */
+  constructor(win, frame) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!HTMLIFrameElement} */
+    this.frame_ = frame;
+
+    /** @private {boolean} */
+    this.isReady_ = false;
+
+    /** @private
+     *  {!../../../src/3p-analytics-common.AmpAnalytics3pEventMap}
+     */
+    this.transportIdToPendingMessages_ = {};
+
+    /** @private {string} */
+    this.messageType_ = AMP_ANALYTICS_3P_EVENT_MESSAGES_TYPE;
+
+    /** @private {!../../../src/iframe-helper.SubscriptionApi} */
+    this.postMessageApi_ = new SubscriptionApi(this.frame_,
+        this.messageType_,
+        true,
+        () => {
+          this.setIsReady();
+        });
+  }
+
+  /**
+   * Returns whether the queue has been marked as ready yet
+   * @return {boolean}
+   * @VisibleForTesting
+   */
+  isReady() {
+    return this.isReady_;
+  }
+
+  /**
+   * Indicate that a cross-domain frame is ready to receive messages, and
+   * send all messages that were previously queued for it.
+   * @VisibleForTesting
+   */
+  setIsReady() {
+    this.isReady_ = true;
+    this.flushQueue_();
+  }
+
+  /**
+   * Returns how many transportId -> message(s) mappings there are
+   * @return {number}
+   * @VisibleForTesting
+   */
+  queueSize() {
+    return Object.keys(this.transportIdToPendingMessages_).length;
```

Comment:
it is a bit controversial that the queue is actually a map. and this method is returning the number of transports, rather than the number of messages.

are we talking about a message queue or transport queue?